File: sorting/quicksort_recursive/quicksort.py
def partition(elements, start, end):
    value = elements[end]
    i = start - 1

    for p in range(start, end):
        if elements[p] <= value:
            i += 1
            elements[i], elements[p] = elements[p], elements[i]

    i += 1
    elements[i], elements[end] = elements[end], elements[i]

    return i

def quicksort(elements, start = 0, size = None):
    if size is None:
        size = len(elements) - 1

    if size == 0 or start >= size:
        return

    p = partition(elements, start, size)
    quicksort(elements, start, p - 1)
    quicksort(elements, p + 1, size)

a = list(reversed(range(11)))
quicksort(a)

print(a)

File: sorting/quicksort_recursive/quicksort.cpp
#include <utility>

#include <iostream>

template <typename T>
std::size_t partition(T *elements, std::size_t start, std::size_t end) {
    T *value = elements + end;
    std::size_t i = start - 1;

    for (auto p = start; p < end; p++) {
        if (elements[p] <= *value) {
            i++;
            std::swap(elements[i], elements[p]);
        }
    }

    i++;
    std::swap(elements[i], elements[end]);

    return i;
}

template <typename T>
void quicksort(T *elements, std::size_t start, std::size_t end) {
    if (elements == nullptr) {
        return;
    }

    if (start >= end) {
        return;
    }

    auto p = partition(elements, start, end);

    if (p != start) {
        quicksort(elements, start, p - 1);
    }

    if (p != end) {
        quicksort(elements, p + 1, end);
    }
}

template <typename T>
void quicksort(T *elements, std::size_t size) {
    if (elements == nullptr) {
        return;
    }

    quicksort(elements, 0, size - 1);
}

int main(void) {
    int elements[] = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0};

    quicksort(elements, 10);

    for (auto v : elements) {
        std::cout << v << " ";
    }
    std::cout << std::endl;

    return 0;
}
